イベント編集ページに基本情報（競技名、都道府県、対象年齢、イベント名、URL、開催場所、性別）を表示

toshinori-m opened the pull request:
## 概要
- イベント編集ページ（EventSettingEditPage.tsx）を新規作成
- 下の項目をAPIから取得してフォームに表示する機能を実装（表示のみ、編集・送信は未対応）

## 表示項目
- 競技名（sportsType）
- 都道府県（prefecture）
- 対象年齢（targetAge）
- イベント名
- イベントURL
- 開催場所（area）
- 性別（sex）

## 備考
- 下の項目の追加及びこのページの編集・送信機能は次回PRで対応予定
  - 「開始日付」「終了日付」「募集チーム数」「イベント目的」「その他」

close https://github.com/toshinori-m/stay_connect/issues/191

toshinori-m committed:
feat: イベント編集ページに基本情報（競技名、都道府県、対象年齢、イベント名、URL、開催場所、性別）を表示

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,271 @@
+                  label="競技名"
+                  value={formState.sportsTypeSelected ? formState.sportsTypeSelected.id : ""}
+                  options={sportsTypes}
+                  onChange={() => {}}
```
まだ未実装箇所だと思うのでTODOコメントがどこかにあると良さそう

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,271 @@
+    }
+  }, [sportsDisciplines])
+
+  const getEventSettingData = async () => {
```
関数や変数名でfetchとgetで表現が揺れているのでapiの取得はfetchで統一したりした方が良さそう
fetchRecruitmentAndSetFormState

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,271 @@
+    }
+  }
+  
+  const getSportsType = async (selectedTypeId?: number) => {
```
api取得の時もgetでformState 更新もgetだと同じgetから始まるのに実際の処理内容が一貫されておらず関数名だけ見て何をしているかをパッと判断できないので例えば

apiでのデータ取得はfetch
stateの更新はselectで始まるようにするなどの一貫性が見えると読みやすくなりそうな気がしました！
selectSportsTypeById

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,271 @@
+import useFetchDisciplines from "@/hooks/search/useFetchDisciplines"
+
+export default function EventSettingForm() {
+  const { id } = useParams<{ id: string }>()
```
なんのidかを明確にした方が良さそう
hoge_id

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,271 @@
+  const { sportsDisciplines, errors: disciplineErrors } = useFetchDisciplines(formState.sportsTypeSelected)
+  const disciplineIdsRef = useRef<number[]>([])
+  const [errors, setErrors] = useState<string[]>([])
+  const hasFetchedRef = useRef(false)
```
再レンダリング等の理由でuseStateではなくuseRefにしたのだとおもますが、useStateで表現できないでしょうか？？
ライフサイクルがわかりにくいのでuseRefは極力使いたくないなと思い

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,271 @@
+        prefectureSelected: recruitmentData.prefecture_id ? recruitmentData.prefecture_id.toString() : null
+      }))
+
+      const rdsRes = await apiClient.get(`/recruitments/${id}/sports_disciplines`)
```
rdsResはなんの略でしょうか？？
パッと見ただけで何しめしているかわかる命名にした方が良いと思います

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,271 @@
+      const rdsRes = await apiClient.get(`/recruitments/${id}/sports_disciplines`)
+      disciplineIdsRef.current = rdsRes.data.map((item: { sports_discipline_id: number }) => item.sports_discipline_id)
+
+      const rtaRes = await apiClient.get(`/recruitments/${id}/target_ages`)
```
rtaResを見ただけで何をしめしているかわからないので明快な命名を考えてください

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,271 @@
+    }
+  }
+
+  const renderErrorList = (errors: string[]) => {
```
renderErrorListは命名から何を返すのかが明確でない気がします
(JSX？文字列？配列？)

ErrorListで良いのではないかと思いますがいかがでしょうか？？

どうしてもrenderを使いたいのであれば
renderErrorListElement
とかのようにtsxを返すとこを明確にした方が良さそう

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,271 @@
+    errors: initialErrors
+  } = useInitialFormData()
+
+  const { sportsDisciplines, errors: disciplineErrors } = useFetchDisciplines(formState.sportsTypeSelected)
```
state定義とstate 定義の間にこのあたりの処理を挟んだ理由は何かあったりしますか？？
state定義は一箇所にまとまっていた方が読みやすい気がしてます

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,271 @@
+      }))
+
+      const rdsRes = await apiClient.get(`/recruitments/${id}/sports_disciplines`)
+      disciplineIdsRef.current = rdsRes.data.map((item: { sports_discipline_id: number }) => item.sports_discipline_id)
```
disciplineIdsRefはこの関数内でセットする必要があるデータでしょうか？？
2つのuseEffectを跨いでデータが関わっており、非常に読みにくくしている印象なので、データが必要な箇所で取得するようにした方が良さそうに見えます

ここで必要なデータなのでこの関数内でデータfetchする必要性はなさそう: https://github.com/toshinori-m/stay_connect/blob/86c163051135a3d5e33c154bee6c76f04347fa6b/frontend-react/src/pages/eventPages/EventSettingEditPage.tsx#L50

hirosi1900day reviewed (commented):
一旦ここまで指摘

toshinori-m committed:
refactor: 可読性・保守性向上のためのコード整理

- useRef を useState に置き換え（再レンダリングの考慮が不要なため）
- API レスポンス変数 rdsRes, rtaRes を意味が分かる名前にリネーム
- JSXを返す関数 renderErrorList を ErrorList にリネームし、返却内容を明確化
- state 定義や useRef の宣言を1か所にまとめてコードの見通しを改善
- useEffect 複数をまたいで使用されていた disciplineIdsRef の責務を見直し、用途ごとに整理

toshinori-m commented:
多数のコメントをありがとうございます！

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,290 @@
+  const { sportsDisciplines, errors: sportsDisciplineErrors } = useFetchDisciplines(formState.sportsTypeSelected)
+
+  useEffect(() => {
+    if (hasFetchedRef.current) return
```
こちらもuseRefを使わないと表現できなかった認識であってますか？？

toshinori-m replied:
あー、すみません！

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,290 @@
+      } catch {
+        setErrors(["種目の取得に失敗しました。"])
+      }
+    }
```
ここはuseEffectの中で関数を定義しており、もう一つのuseEffectは外で定義した関数を使う構成になっており統一感がないのでどちらかに統一してみてください

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,290 @@
+    if (hasFetchedRef.current) return
+    if (!recruitmentId || sportsTypes.length === 0 || prefectures.length === 0 || targetAges.length === 0) return
+
+    fetchRecruitmentAndSetFormState()
```
この関数の中で取得とstateのsetを行ってますが、その処理を分離するした方が単一責務の原則に従うのでわかりやすくなりそうな気がしてます

戻り値 = fetch関数
set関数(戻り値)

toshinori-m committed:
refactor: 以下の処理をリファクタリング

- useEffect内外の関数定義方法を統一
- fetch処理とstate更新処理を分離

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,311 @@
+    await setSelectedSportsType(recruitmentData.sports_type_id)
+    await setSelectedTargetAge(targetAgeIds)
+
+    setHasFetched(true)
```
setHasFetched(false)の処理も書いてやらないとおかしな動きはしないですかね？？
動作確認が取れているのであれば大丈夫ですが

toshinori-m replied:
ありがとうございます！

`setHasFetched` の代わりに `fetchedId` を使って、`recruitmentId` ごとにフェッチ済みかを判定するようにしました。
これで編集ページを再利用したり、IDが切り替わっても対応できるようになっていると考えています！

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,311 @@
+
+    await setRecruitmentFormState(recruitmentData)
+    await setSelectedSportsType(recruitmentData.sports_type_id)
+    await setSelectedTargetAge(targetAgeIds)
```
ここを非同期にする理由はありますでしょうか？？

toshinori-m committed:
Refactor: setHasFetched を廃止し、fetchedId による状態を管理するように変更

toshinori-m committed:
refactor: setFormState 関連処理から不要な await を削除

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,312 @@
+    setFetchedId(recruitmentId ?? null)
+  }
+
+  const fetchRecruitmentData = async () => {
```
fetchAndInitializeFormは命名的に２つの責務を持っているので、
この関数をかまさずにuseEffectの中でfetchに成功したらstate setみたいにした方が読みやすいコードになるかなと思いますがいかがでしょうか？

例えば以下みたいなイメージ
```tsx
useEffect(() => {
  if (fetchedId === recruitmentId) return
  if (sportsTypes.length === 0 || prefectures.length === 0 || targetAges.length === 0) return
  if (!recruitmentId) return

  fetchRecruitmentData(recruitmentId)
    .then(recruitmentFormData => {
      if (!recruitmentFormData) return

      const { recruitmentData, targetAgeIds } = recruitmentFormData

      setRecruitmentFormState(recruitmentData)
      setSelectedSportsType(recruitmentData.sports_type_id)
      setSelectedTargetAge(targetAgeIds)
      setFetchedId(recruitmentId)
    })
    .catch(() => {
      setErrors(["イベントを表示できませんでした。"])
    })
  // eslint-disable-next-line react-hooks/exhaustive-deps
}, [recruitmentId, sportsTypes, prefectures, targetAges])
```

toshinori-m replied:
コメントありがとうございます！

hirosi1900day reviewed (commented):
すみません
最後どうしても気になった箇所コメントしました🙏

toshinori-m committed:
refactor: fetchAndInitializeFormを削除し、useEffect内で処理を分離して可読性を改善

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,313 @@
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [sportsDisciplines])
+
+  const fetchAndSetDisciplines = async () => {
```
ここもfetchとsetの２つの責務を持って意いるのでthenでの対応をすると読みやすくなると思います

toshinori-m replied:
やはり気づかれていましたね！(⌒-⌒; )

toshinori-m committed:
refactor: fetchとsetの責務を分離し、useEffect内で処理を明確化

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,313 @@
+    }))
+  }
+
+  useEffect(() => {
```
setSelectedTargetAge と setSelectedSportsType は48行目の useEffect 内で使われているので、useEffect をこの位置に定義するのではなく、これらの関数の後に定義したほうが、関数の使われる位置との距離が近くなり、読みやすくなると思います。
（関数やフックは「使われる場所の近くに定義する」方がコードの流れが追いやすくなります）

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,313 @@
+  
+
+  const setSelectedSportsType = async (selectSportsTypeId?: number) => {
+    try {
```
エラーは発生しないのでtryは不要(混乱の元になるため)

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,313 @@
+
+  const setSelectedSportsType = async (selectSportsTypeId?: number) => {
+    try {
+      setErrors([])
```
エラーは発生しない

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,313 @@
+        }
+      }
+    } catch {
+      setErrors(["競技を表示できませんでした。"])
```
エラーは発生しない

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,313 @@
+  }
+
+  const setSelectedTargetAge = async (targetAgeIds?: number[]) => {
+    try {
```
エラーは発生しない(api通信はないので)

hirosi1900day commented on frontend-react/src/pages/eventPages/EventSettingEditPage.tsx:
```diff
@@ -0,0 +1,313 @@
+        }))
+      }
+    } catch {
+      setErrors(["対象年齢を表示できませんでした。"])
```
いらない

toshinori-m committed:
Refactor: 不要な async/try-catch を削除し、エラー初期化の位置を適正化
